### PR TITLE
Added ButtonStyle.maximumSize

### PR DESCRIPTION
**Pull request by HansMuller**
Added a ButtonStyle maximumSize property. Its default value is Size.infinite for TextButton, ElevatedButton, OutlinedButton, i.e. none of the buttons constrain their maximum size by default.

### CODE REVIEW

File: packages/flutter/lib/src/material/button_style.dart [hunk 1]
```diff
@@ -159,18 +160,29 @@ class ButtonStyle with Diagnosticable {
   ///
   /// The size of the rectangle the button lies within may be larger
   /// per [tapTargetSize].
+  ///
+  /// This value must be less than or equal to [maximumSize].
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/button_style.dart:164]: Where is this enforced?

File: packages/flutter/lib/src/material/button_style.dart [hunk 2]
```diff
@@ -159,18 +160,29 @@ class ButtonStyle with Diagnosticable {
...
   /// `fixedSize: Size.fromWidth(320)`. Similarly, to specify a fixed
   /// height and the default width use `fixedSize: Size.fromHeight(100)`.
   final MaterialStateProperty<Size?>? fixedSize;
 
+  /// The maximum size of the button itself.
+  ///
+  /// A [Size.infinite] or null value for this property means that
+  /// the button's maximum size is not constrained.
+  ///
+  /// This value must be greater than or equal to [minimumSize].
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/button_style.dart:183]: Where is this enforced?

**Review by darrenaustin — APPROVED** [review]
LGTM.

I probably missed it, but I didn't see where the min < max size constraint is enforced. Is this handled by the `BoxContraints`?

File: packages/flutter/lib/src/material/button_style.dart (hunk 1 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style.dart:164]: In the _ButtonStyleStat.build() expression that builds a BoxConstraints object:
```dart
    BoxConstraints effectiveConstraints = resolvedVisualDensity.effectiveConstraints(
      BoxConstraints(
        minWidth: resolvedMinimumSize!.width,
        minHeight: resolvedMinimumSize.height,
        maxWidth: resolvedMaximumSize!.width,
        maxHeight: resolvedMaximumSize.height,
      ),
    );
```

If you pass invalid minimum/maximumSize values you get a pretty obvious error from the BoxConstraints constructor.

File: packages/flutter/lib/src/material/button_style.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style.dart:183]: See above.

**Pull request merged**